src/crimson/osd/watch.h

```diff
@@ -0,0 +1,187 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <iterator>
+//#include <list>
+#include <memory>
+#include <map>
+#include <set>
+
+#include "crimson/net/Connection.h"
+#include "crimson/osd/object_context.h"
+
+namespace crimson::osd {
+
+class Notify;
+using NotifyRef = std::shared_ptr<Notify>;
+
+// NOTE: really need to have this public. Otherwise `shared_from_this()`
+// will abort. According to cppreference.com:
+//
+//   "The constructors of std::shared_ptr detect the presence
+//   of an unambiguous and accessible (ie. public inheritance
+//   is mandatory) (since C++17) enable_shared_from_this base".
+class Watch : public std::enable_shared_from_this<Watch> {
+  struct NotifyCmp {
+    inline bool operator()(NotifyRef lhs, NotifyRef rhs) const;
+  };
+  ::std::set<NotifyRef, NotifyCmp> in_progress_notifies;
+  ::crimson::net::ConnectionRef conn;
+  ::crimson::osd::ObjectContextRef obc;
+
+  watch_info_t winfo;
+  entity_name_t entity_name;
+
+  seastar::future<> start_notify(NotifyRef);
+  seastar::future<> send_notify_msg(NotifyRef);
+  seastar::future<> send_disconnect_msg();
+  void discard_state();
+
+  friend Notify;
+public:
+  // XXX: hack for make_shared
+  Watch(::crimson::osd::ObjectContextRef obc,
+        const watch_info_t& winfo,
+        const entity_name_t& entity_name)
+    : obc(std::move(obc)),
+      winfo(winfo),
+      entity_name(entity_name) {
+  }
+
+public:
+  seastar::future<> connect(crimson::net::ConnectionRef, bool);
+  bool is_alive() const {
+    return true;
+  }
+  bool is_connected() const {
+    return static_cast<bool>(conn);
+  }
+  void got_ping(utime_t) {
+    // NOP
+  }
+
+  seastar::future<> remove(bool);
+
+  /// Call when notify_ack received on notify_id
+  seastar::future<> notify_ack(
+    uint64_t notify_id, ///< [in] id of acked notify
+    const ceph::bufferlist& reply_bl); ///< [in] notify reply buffer
+
+  // TODO: we don't need the atomical ref-counting but shared_from_this
+  // will be useful. Maybe switch to something from boost later.
+  using Ref = std::shared_ptr<Watch>;
```

Comment:
I assumed watch / notify are not supposed to participate in hot paths.

Reply:
Switching to `seastar::shared_ptr`. `lw_shared_ptr` would inconvenient because `Watch` in `object_context.h` is only forward-declared.